rasa/graph_components/validators/config_validator.py

```diff
@@ -0,0 +1,532 @@
+                f"You have defined multiple entity extractors that do the same job "
+                f"in your pipeline: "
+                f"{types_to_str(extractors_in_pipeline)}. "
+                f"This can lead to the same entity getting "
+                f"extracted multiple times. Please read the documentation section "
+                f"on entity extractors to make sure you understand the implications: "
+                f"{DOCS_URL_COMPONENTS}#entity-extractors"
+            )
+
+    def _warn_of_competition_with_regex_extractor(
+        self, training_data: TrainingData
+    ) -> None:
+        """Warns when regex entity extractor is competing with a general one.
+
+        This might be the case when the following conditions are all met:
+        * You are using a general entity extractor and the `RegexEntityExtractor`
+        * AND you have regex patterns for entity type A
+        * AND you have annotated text examples for entity type A
+
+        Args:
+            training_data: the training data
+        """
+        present_general_extractors = self._component_types.intersection(
+            TRAINABLE_EXTRACTORS
+        )
+        has_general_extractors = len(present_general_extractors) > 0
+        has_regex_extractor = (
+            RegexEntityExtractorGraphComponent in self._component_types
+        )
+
+        regex_entity_types = {rf["name"] for rf in training_data.regex_features}
+        overlap_between_types = training_data.entities.intersection(regex_entity_types)
+        has_overlap = len(overlap_between_types) > 0
+
+        if has_general_extractors and has_regex_extractor and has_overlap:
+            rasa.shared.utils.io.raise_warning(
+                f"You have an overlap between the "
+                f"{RegexEntityExtractorGraphComponent.__name__} and the "
+                f"statistical entity extractors "
+                f"{types_to_str(present_general_extractors)} "
+                f"in your pipeline. Specifically both types of extractors will "
+                f"attempt to extract entities of the types "
+                f"{', '.join(overlap_between_types)}. "
+                f"This can lead to multiple "
+                f"extraction of entities. Please read "
+                f"{RegexEntityExtractorGraphComponent.__name__}'s "
+                f"documentation section to make sure you understand the "
+                f"implications: {DOCS_URL_COMPONENTS}#regexentityextractor"
+            )
+
+    def _raise_if_featurizers_are_not_compatible(self,) -> None:
+        """Raises or warns if there are problems regarding the featurizers.
+
+        Raises:
+            `InvalidConfigException`
+        """
+        featurizers: List[SchemaNode] = [
+            node
+            for node in self._graph_schema.nodes.values()
+            if issubclass(node.uses, Featurizer2)
+        ]
+        Featurizer2.raise_if_featurizer_configs_are_not_compatible(
```

Comment:
how about using protocols for this? That would avoid that users / we have to implement subclasses

Reply:
This is a class method that tests if the names of the featurizers overlap which would cause trouble downstream. I'm not sure where protocols would be useful here?

Reply:
And in general I think it would be cool if we could refactor the base classes such that if we use e.g. a Featurizer we know / it is guaranteed that <insert some feature logic here> is done to a message. That would avoid confusion around what featurizer need to do to make them compatible with everything else (which is why we have the strange lexical syntactic featurizer issue now).

Reply:
> I'm not sure where protocols would be useful here?

Yeah, sorry. Completely misread that one 🤦🏻 Is it weird that all the checks are part of the ConfigValidator but this one check is part of the `Featurizer` class? Should we move it into `ConfigValidator`?

I think I need some more explanations around the last point 🙈 Which lexical featurizer issues are you referring to?

Reply:
> one check is part of the Featurizer class?
Because that check requires understanding why/when featurizers are incompatible, I thought it might be better to keep that logic closer to the Featurizers. Do we need to make that visible here?
>  lexical featurizer issues
that was the one not producing sentence features and the only reason why the "add features to message" function had mandatory sequence but optional sentence features :D - though I'm not sure there's a good way to enforce that through interface design and whether there is much code to be shared - maybe protocols + tests are the way to make those assumptions explicit 🤔

Reply:
> Do we need to make that visible here?

Both is fine - it's rather a nit pick as the implementation currently is very high level and only check whether we have some duplicate aliases 🤷🏻 

I think the code sharing through the interfaces makes us quite inflexible because although a lot of stuff is similar, some components are always different.